src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,700 @@
+        }
+        copy_op.dst_extent_map[dst_offset] = e.second;
+        src_offset += e.second;
+        dst_offset += e.second;
+      }
+      if (dst_offset < copy_op.dst_offset + copy_op.length) {
+        uint64_t zero_len = copy_op.dst_offset + copy_op.length - dst_offset;
+        ldout(m_cct, 20) << "src_snap_seq=" << src_snap_seq
+                         << ", inserting zero " << dst_offset << "~"
+                         << zero_len << dendl;
+        m_dst_zero_interval[src_snap_seq].insert(dst_offset, zero_len);
+      } else {
+        assert(dst_offset == copy_op.dst_offset + copy_op.length);
+      }
+      m_write_ops[src_snap_seq].emplace_back(std::move(copy_op));
+    }
+  }
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::compute_zero_ops() {
+  ldout(m_cct, 20) << dendl;
+
+  bool fast_diff;
+  {
+    RWLock::RLocker snap_locker(m_dst_image_ctx->snap_lock);
+    fast_diff = m_dst_image_ctx->test_features(RBD_FEATURE_FAST_DIFF,
+                                               m_dst_image_ctx->snap_lock);
+  }
+  bool prev_exists = false;
+  uint64_t prev_end_size = 0;
+  for (auto &it : m_dst_zero_interval) {
+    auto src_snap_seq = it.first;
+    auto &zero_interval = it.second;
+
+    ldout(m_cct, 20) << "src_snap_seq=" << src_snap_seq << ", prev_end_size="
+                     << prev_end_size << dendl;
+
+    uint64_t end_size = m_dst_object_end[src_snap_seq];
+
+    if (zero_interval.empty()) {
+      end_size = std::max(prev_end_size, end_size);
+    }
+    for (auto z = zero_interval.begin(); z != zero_interval.end(); z++) {
+      if (z.get_start() + z.get_len() == m_dst_object_end[src_snap_seq]) {
+        // zero interval at the object end
+        if (z.get_start() < prev_end_size) {
+          if (z.get_start() == 0) {
+            m_write_ops[src_snap_seq]
+              .emplace_back(COPY_OP_TYPE_REMOVE, 0, 0, 0);
+            ldout(m_cct, 20) << "COPY_OP_TYPE_REMOVE" << dendl;
+          } else {
+            m_write_ops[src_snap_seq]
+              .emplace_back(COPY_OP_TYPE_TRUNC, 0, z.get_start(), 0);
+            ldout(m_cct, 20) << "COPY_OP_TYPE_TRUNC " << z.get_start() << dendl;
+          }
+        }
+        end_size = z.get_start();
+      } else {
+        // zero interval inside the object
+        assert(z.get_start() + z.get_len() < m_dst_object_end[src_snap_seq]);
+        m_write_ops[src_snap_seq]
+          .emplace_back(COPY_OP_TYPE_ZERO, 0, z.get_start(), z.get_len());
+        ldout(m_cct, 20) << "COPY_OP_TYPE_ZERO " << z.get_start() << "~"
+                         << z.get_len() << dendl;
+      }
+    }
+    ldout(m_cct, 20) << "end_size=" << end_size << dendl;
+    if (end_size == 0) {
+      m_dst_object_state.erase(src_snap_seq);
+      prev_exists = false;
+    } else {
+      if (fast_diff && prev_exists && m_write_ops[src_snap_seq].empty() &&
+          end_size == prev_end_size) {
+        m_dst_object_state[src_snap_seq] = OBJECT_EXISTS_CLEAN;
```

Comment:
The state should also be clean if you don't have any write ops and the object existed (clean or dirty) in the previous destination snapshot

Reply:
@dillaman I am not sure I see how it is different from what I have. Anyway, I have made some cleanup here (I realized `prev_exists` var is not needed as I can use `prev_end_size > 0` as a test). Do you see uncovered cases with the current version?